### ISSUE TITLE
Traverse expander breaks on IP addresses close to zero

### ISSUE THREAD

**Issue by timmc** [mschwager/fierce#29]
The traverse expander subtracts small numbers from an IP address and constructs new IPs from the results. This breaks when the result goes negative, e.g. traversing backwards from `0.0.0.0`.

I had a little trouble reproducing this, but if you have a subdomain that is a CNAME of a domain that resolves to `0.0.0.0`, you'll get a stack trace like this:

```
Traceback (most recent call last):
  File "./fierce/fierce.py", line 486, in <module>
    main()
  File "./fierce/fierce.py", line 480, in main
    fierce(**vars(args))
  File "./fierce/fierce.py", line 373, in fierce
    ips = expander_func(ip)
  File "./fierce/fierce.py", line 173, in traverse_expander
    result = [ipaddress.IPv4Address(ip + i) for i in range(-n, n + 1)]
  File "./fierce/fierce.py", line 173, in <listcomp>
    result = [ipaddress.IPv4Address(ip + i) for i in range(-n, n + 1)]
  File "/usr/lib/python3.5/ipaddress.py", line 575, in __add__
    return self.__class__(int(self) + other)
  File "/usr/lib/python3.5/ipaddress.py", line 1269, in __init__
    self._check_int_address(address)
  File "/usr/lib/python3.5/ipaddress.py", line 420, in _check_int_address
    raise AddressValueError(msg % (address, self._version))
ipaddress.AddressValueError: -5 (< 0) is not permitted as an IPv4 address
```

I don't have a public domain I can share for this, but if you CNAME something to `qa.cloudapp.net` it should do the trick. :-)

**Comment by timmc** [comment]
I have a fix locally along with some tests:

```python
MAX_IP = int(ipaddress.IPv4Address('255.255.255.255'))

def traverse_expander(ip, n=5):
    class_c = get_class_c_network(ip)

    ip_min = max(int(ip) - n, 0)
    ip_max = min(int(ip) + n, MAX_IP)
    result = [ipaddress.IPv4Address(i) for i in range(ip_min, ip_max + 1)]
    result = [i for i in result if i in class_c]

    return result
```

...but it occurs to me that the class-bounding is a little redundant with the other min/max stuff and they could be combined. However, `IPv4Address` doesn't seem to have any convenient "take this apart into octets" so I'll need to poke at this a bit longer.

**Comment by mschwager** [comment]
Your fix looks sane to me! One suggestion I have would be to just constant-ify the max IP - using `ipaddress.IPv4Address` seems like unnecessary overhead. Something like:

```python
def traverse_expander(ip, n=5):
    class_c = get_class_c_network(ip)

    ip = int(ip)
    ipv4_min = 0
    ipv4_max = 2**32 - 1
    ip_min = max(ip - n, ipv4_min)
    ip_max = min(ip + n, ipv4_max)

    result = [ipaddress.IPv4Address(i) for i in range(ip_min, ip_max + 1)]
    result = [i for i in result if i in class_c]

    return result
```

> it occurs to me that the class-bounding is a little redundant with the other min/max stuff and they could be combined.

Not entirely sure what you're getting at here, but this solution LGTM!

**Comment by timmc** [comment]
Re: the class-bounding, this is probably just bikeshedding, but I was noticing that the filtering IPs by whether they're in the same class C could be accomplished at the same time as the bounds check filtering by just setting the ip_min and ip_max to the lowest and highest of the class C (rather than the lowest and highest of *all* IPv4 addresses.) A little math oughtta do it. :-)

**Comment by mschwager** [comment]
Heads up @timmc, I just released version `1.3.0` (https://pypi.org/project/fierce/1.3.0/) which includes your changes. Thanks again!
